proposals/base-relative-assets.md

```diff
@@ -0,0 +1,132 @@
+  build: {
+    linkStyle: 'absolute'
+  }
+});
+```
+
+Your CSS will build to this:
+
+```html
+<link rel="stylesheet" href="/docs/assets/main.hash.css">
+```
+
+Using the `relative` config option:
+
+```js
+import { defineConfig } from 'astro/config';
+
+export default defineConfig({
+  build: {
+    linkStyle: 'relative'
+  }
+});
+```
+
+You'll instead get this:
+
+```html
+<link rel="stylesheet" href="../assets/main.hash.css">
+```
+
+# Motivation
+
+We often get support questions on how to deploy an Astro site without knowing the base in advance. There has been discussions and RFCs to solve this problem such as [this one](https://github.com/withastro/rfcs/discussions/334) and [this one](https://github.com/withastro/rfcs/discussions/23).
+
+People have tried to use `./` in the `base` option to make this happen to, such as in [this issue](https://github.com/withastro/astro/issues/4229).
+
+The primary use-case is wanting to deploy a site that might live under multiple subdomains. Using relative links makes it portable.
+
+# Detailed design
+
+This new configuration option will affect these types of assets:
+
+- Link tags added by us from bundled CSS.
+- Hoisted script tags added by us from bundled hoisted scripts.
+- Links to scripts added to `<astro-island>`s.
+
+Any other type of asset, such as images, are not controlled by Astro and will continue to not be so. The developer will need to make these links relative themselves. They can also use the `<base>` tag to make this easier.
+
+## Implementation
+
+During the build when we go to generate pages there are a few places where we resolve the final URL for links.
+
+- For link tags it is done [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L300).
+- For hoisted scripts it is [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L301)
+- For links added to astro-islands it [is here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L352).
+
+We should generalize this logic so that there's one function that gets called to resolve the URL given the `linkStyle` option, as well as the base and subpath. That function will use the `linkStyle` config to determine which algorithm to use.
+
+For SSR the same things are done [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/app/index.ts#L63) for example. Again, this should be consolidated so we're not doing the same sort of logic in many different places, creating areas for bugs. Instead SSG and SSR should use shared logic.
+
+## 404.astro
+
+If using `linkStyle: 'relative'` a 404.astro page __will not be allowed__. This is because it is impossible to make these links be relative in SSG mode. That's because a 404.astro might be used by paths such as:
```

Comment:
I've think long and hard about this but I can't seem to find a nice solution too. Unless we get hybrid support soon maybe then it could work as a middle ground.

Reply:
And just as a sanity check: `<base>` can't help us here right? Or is it that the user has to manually add the base tag themselves, which is a footgun.

```
  <base href="/">
```